client/src/main/java/com/google/cloud/pubsub/AbstractSubscriberConnection.java

```diff
@@ -310,7 +310,10 @@ public void run() {
 
   private void addOutstadingAckHandlers(
       ExpirationInfo expiration, final List<AckHandler> ackHandlers) {
-    outstandingAckHandlers.putIfAbsent(expiration, new ArrayList<AckHandler>(ackHandlers.size()));
```

Comment:
It is expected to be atomic, you might want to check the library in https://github.com/GoogleCloudPlatform/google-cloud-java/tree/master/google-cloud-pubsub

It is based on this code and will replace it.